File: javascript/doubly_linked_list.js
class DoublyNode
{
    constructor(data)
    {
        this.data = data; // Value of the node
        this.next = null; // Pointer to the next node
        this.prev = null; // Pointer to the previous node
    }
}

class DoublyLinkedList
{
    constructor()
    {
        this.head = null; // Start of the list
        this.tail = null; // End of the list
        this.size = 0; // Total number of nodes
    }

    // Append: Add a node at the end
    append(data)
    {
        const newNode = new DoublyNode(data); // Create a new node
        if (!this.head)
        {
            // If the list is empty, set head and tail to the new node
            this.head = this.tail = newNode;
        }
        else
        {
            this.tail.next = newNode; // Link the current tail to the new node
            newNode.prev = this.tail; // Link the new node back to the current tail
            this.tail = newNode; // Update the tail to the new node
        }
        this.size++; // Increment the size of the list
    }

    // Prepend: Add a node at the beginning
    prepend(data)
    {
        const newNode = new DoublyNode(data); // Create a new node
        if (!this.head)
        {
            // If the list is empty, set head and tail to the new node
            this.head = this.tail = newNode;
        }
        else
        {
            newNode.next = this.head; // Link the new node to the current head
            this.head.prev = newNode; // Link the current head back to the new node
            this.head = newNode; // Update the head to the new node
        }
        this.size++; // Increment the size of the list
    }

    // Display Forward: Display elements from head to tail
    displayForward()
    {
        let current = this.head;
        console.log("List (Forward):");
        while (current)
        {
            console.log(current.data);
            current = current.next;
        }
    }

    // Display Backward: Display elements from tail to head
    displayBackward()
    {
        let current = this.tail;
        console.log("List (Backward):");
        while (current)
        {
            console.log(current.data);
            current = current.prev;
        }
    }

    // Get Size: Return the number of nodes in the list
    getSize()
    {
        return this.size;
    }

    // Search: Check if a value exists in the list
    search(data)
    {
        let current = this.head;
        while (current)
        {
            if (current.data === data)
            {
                return true; // Value found
            }
            current = current.next;
        }
        return false; // Value not found
    }

    // Remove: Delete the first occurence of a value
    remove(data)
    {
        if (!this.head) return; // If the list is empty, do nothing

        if (this.head.data === data)
        {
            // If the head is the node to remove
            this.head = this.head.next; // Update the head
            if (this.head) this.head.prev = null; // Clear the previous pointer
            else this.tail = null; // If the list is now empty, clear the tail
            this.size--;
            return;
        }
        let current = this.head;

        while (current && current.data !== data)
        {
            current = current.next;
        }

        if (current)
        {
            // Node found
            if (current.next) current.next.prev = current.prev; // Update next node's pointer
            else this.tail = current.prev; // If it's the tail, update the tail
            if (current.prev) current.prev.next = current.next; // Update previous node's next pointer
            this.size--;
        }
    }

    // Insert: Add a ndoe at a specific index
    insertAt(index, data)
    {
        if (index < 0 || index > this.size) return; // Invalid index

        const newNode = new DoublyNode(data);

        if (index === 0)
        {
            // Insert at the beginning
            this.prepend(data);
        }
        else if (index === this.size)
        {
            // Insert at the end
            this.append(data);
        }
        else
        {
            let current = this.head;
            let count = 0;

            while (count < index)
            {
                current = current.next;
                count++;
            }
            newNode.next = current; // Link new node to current node
            newNode.prev = current.prev; // Link new node back to previous node
            current.prev.next = newNode; // Update previous node's next pointer
            current.prev = newNode; // Update current node's previous pointer
            this.size++;
        }
    }

    // Reverse: Reverse the order of the list
    reverse()
    {
        let current = this.head;
        let temp = null;
        while (current)
        {
            temp = current.prev; // Swap previous and next pointers
            current.prev = current.next;
            current.next = temp;
            current = current.prev; // Move to the next node (which is now the previous node)
        }

        if (temp)
        {
            this.head = temp.prev; // Update the head to the new front
        }
    }
}

const list = new DoublyLinkedList();
list.append(10);
list.append(20);
list.append(30);
list.prepend(5); // List is now: 5 -> 10 -> 20 -> 30
list.displayForward();
list.displayBackward();
console.log("Size of the list:", list.getSize()); // Output: 4
console.log("Is 20 in the list?", list.search(20)); // Output: true
list.remove(20); // List is now: 5 -> 10 -> 30
list.insertAt(1, 15); // List is now: 5 -> 15 -> 10 -> 30
list.reverse(); // List is now reversed
list.displayForward(); // Prints reversed list

File: kotlin/src/DoublyLinkedList.kt
class DoublyNode(var value: Int)
{
    var prev: DoublyNode? = null // Pointer to the previous node
    var next: DoublyNode? = null // Pointer to the next node
}

class DoublyLinkedList
{
    private var head: DoublyNode? = null // Head of the list
    private var tail: DoublyNode? = null // Tail of the list

    fun insertAtHead(value: Int)
    {
        val newNode = DoublyNode(value)
        if (head == null)
        {
            head = newNode
            tail = newNode
        }
        else
        {
            newNode.next = head
            head?.prev = newNode
            head = newNode
        }
        println("Inserted $value at head.")
    }

    fun insertAtTail(value: Int)
    {
        val newNode = DoublyNode(value)
        if (tail == null)
        {
            head = newNode
            tail = newNode
        }
        else
        {
            newNode.prev = tail
            tail?.next = newNode
            tail = newNode
        }
        println("Inserted $value at tail.")
    }

    fun printListForward()
    {
        println("List (Head to Tail):")
        var current = head
        while (current != null)
        {
            println("${current.value}")
            current = current.next
        }
        println()
    }

    fun printListBackward()
    {
        println("List (Tail to Head):")
        var current = tail
        while (current != null)
        {
            print("${current.value}")
            current = current.prev
        }
        println()
    }

    fun search(value: Int): Boolean
    {
        var current = head
        while (current != null)
        {
            if (current.value == value)
            {
                println("Value $value found in the list.")
                return true
            }
            current = current.next
        }
        println("Value $value not found in the list.")
        return false
    }

    fun deleteNode(value: Int)
    {
        var current = head

        while (current != null)
        {
            if (current.value == value)
            {
                if (current.prev != null)
                {
                    current.prev?.next = current.next
                }
                else
                {
                    head = current.next
                }

                if (current.next != null)
                {
                    current.next?.prev = current.prev
                }
                else
                {
                    tail = current.prev
                }

                println("Deleted node with value $value.")
                return
            }
            current = current.next
        }
        println("Node with value $value not found.")
    }

    fun getSize(): Int
    {
        var size = 0
        var current = head
        while (current != null)
        {
            size++
            current = current.next
        }
        return size
    }

    fun insertAtPosition(position: Int, value: Int)
    {
        if (position < 1)
        {
            println("Invalid position!")
            return
        }

        val newNode = DoublyNode(value)
        if (position == 1)
        {
            insertAtHead(value)
            return
        }

        var current = head
        var index = 1
        while (current != null && index < position - 1)
        {
            current = current.next
            index++
        }

        if (current == null)
        {
            println("Position out of bounds.")
            return
        }

        newNode.next = current.next
        newNode.prev = current
        current.next?.prev = newNode
        current.next = newNode
        println("Inserted $value at position $position.")
    }

    fun reverse()
    {
        var current = head
        var temp: DoublyNode? = null
        while (current != null)
        {
            temp = current.prev
            current.prev = current.next
            current.next = temp
            current = current.prev
        }
        if (temp != null)
        {
            head = temp.prev
        }
        println("Reversed the list.")
    }
}

fun main()
{
    val list = DoublyLinkedList()

    list.insertAtHead(10)
    list.insertAtTail(20)
    list.insertAtTail(30)
    list.insertAtHead(5)
    list.printListForward() // Output: 5 10 20 30

    list.search(10) // Output: Value 10 found in the list.
    list.search(50) // Output: Value 50 not found in the list.

    list.insertAtPosition(3, 15) // Insert 15 at position 3
    list.printListForward() // Output: 5 10 15 20 30

    list.deleteNode(15) // Output: Deleted node with value 15.
    list.printListForward() // Output: 5 10 20 30

    println("Size of the list: ${list.getSize()}") // Output: Size of the list: 4

    list.printListBackward() // Output: 30 20 10 5

    list.reverse()
    list.printListForward() // Output: 30 20 10 5
}